### PR TITLE
JENKINS-33255: Pull largefiles on slave cached repos if master repo requires them

### PR DESCRIPTION
**Pull request by richard-fine**
[JENKINS-33255](https://issues.jenkins-ci.org/browse/JENKINS-33255)

### CODE REVIEW

File: src/main/java/hudson/plugins/mercurial/Cache.java [hunk 1]
```diff
@@ -216,6 +216,15 @@ private synchronized ReentrantLock getLockForSlaveNode(String node) {
                         return null;
                     }
                 }
+				
+				// Fix for JENKINS-33255: if the clone of the repository on the master indicates that it requires largefiles, then 
+				// pull them for the slave's cached repository. When the repository is the cloned/updated into the workspace, the 
+				// clone can then bring in all the largefiles without a problem. Largefiles are shared (hardlinked) on a single machine 
+				// so we will not end up with extra copies.
+				if (masterHg.requiresLargefiles(masterCache)) {
+					slaveHg.lfpull(localCache, null, remote);
```
Comment (richard-fine) [inline src/main/java/hudson/plugins/mercurial/Cache.java]: This is not quite right; we need to pull largefiles for the revision we are actually going to use in the workspace, not just for '.' (which would be tip, and might be a totally different branch to the one we actually want to use).

Comment (richard-fine) [inline src/main/java/hudson/plugins/mercurial/Cache.java]: Fixed in  d79f9b7

**Comment by jglick** [comment]
Probably OK, but there is no test coverage; punting this to a rainy day.

**Comment by jglick** [comment]
Given the extremely limited time I can devote to ongoing maintenance of this plugin, I must unfortunately close PRs without test coverage that could ensure I do not regress something more critical.

**Comment by Artalus** [comment]
This has bitten us quite hard recently. I believe we can manage by manually `hg lfpull https://sources --rev="$REPO_REV"` but it would be needed for every potentially-largefiled repo in every job that relies on caching, which kinda sucks.

@richard-fine @jglick any hint what should be tested for this PR to be merged into one of the next updates? I am not a Java expert, but I would very much like for this bug to be fixed, even if it means I'll have to do it on my own.

**Comment by jglick** [comment]
Acc. to [this](https://www.mercurial-scm.org/wiki/LargefilesExtension) the extension is distributed by default so there should be no need to edit `MercurialContainer/Dockerfile` (except perhaps to refresh the versions listed there to the latest release in each line). Then look at `PipelineTest` or `FunctionalTest` to see how to write a new test class which would create an agent, run Hg 4.x in `MercurialContainer`, turn on caching, create a repository using the extension, run a build using that repository on the agent, and verify that it fails in the expected way.

I have to warn you that this all may not work at all as things stand, since the test infrastructure in the plugin is not really up to it. I have not yet set things up to allow test Hg repositories to be hosted in distinct Docker repositories using the actual Hg server (as opposed to local file protocol), perhaps using [Testcontainers](https://www.testcontainers.org/modules/docker_compose/). Worse, `JenkinsRule` itself does not support running the Jenkins master in a container either, so anyone running the test would need to have a sufficiently new copy of Hg installed locally, which ci.jenkins.io machines will not; and the caching system (as well as multibranch support) currently relies on running Hg commands from the master.

> I am not a Java expert

Indeed. From an economic perspective it is undesirable to have “long-tail” features like Mercurial support (sorry but true) driven by in-JVM plugins. I [prototyped a system](https://github.com/jglick/jk--) (implementing a subset of this plugin’s functionality as a demo, as it happens) that would be much easier for developers without specialized Java/Jenkins expertise to contribute to, as well as offering a much more flexible and relaxed deployment model than pushing a single plugin update to the world and hoping that regressions in corner cases would not force other users to roll back. Unfortunately I have not (yet?) gotten interest from my employer to push this work forward.

**Pull request closed without merging**